File: client/src/session/session.ts
import { b64d, b64e, fromUtf8 } from "../protocol/bytes";
import {
  authMessage,
  bleepCode,
  freshX,
  generateIdentity,
  parseBleepCode,
  REGION,
  registrationMessage,
  sign,
  spkMessage,
  type Identity,
} from "../protocol/keys";
import {
  dumpRatchet,
  encodePt,
  initAlice,
  initBob,
  loadRatchet,
  ratchetDecrypt,
  ratchetEncrypt,
  x3dhAlice,
  x3dhBob,
  type RatchetState,
} from "../protocol/ratchet";
import { decodeInner, encodeInner, openOuter, sealOuter, signInner, verifyInner } from "../protocol/seal";
import { safetyNumber } from "../protocol/safety";
import { api } from "../net/api";
import type { Message, Peer, SessionRec, Thread, VaultSnapshot } from "../types";
import { inspectDraft, type Hit } from "../guardian/engine";

export type SpkHold = { id: number; sk: Uint8Array; pk: Uint8Array };
export type OpkHold = { id: number; sk: Uint8Array; pk: Uint8Array };

export async function registerOnRelay(id: Identity, spk: SpkHold, opks: OpkHold[]) {
  const body = {
    mailbox_id: id.mailboxId,
    identity_ed25519_b64: b64e(id.edPk),
    identity_x25519_b64: b64e(id.xPk),
    signed_prekey: {
      key_id: spk.id,
      pub_b64: b64e(spk.pk),
      sig_b64: b64e(sign(id.edSk, spkMessage(spk.id, spk.pk))),
    },
    one_time_prekeys: opks.map((k) => ({ key_id: k.id, pub_b64: b64e(k.pk) })),
    registration_sig_b64: b64e(sign(id.edSk, registrationMessage(id.mailboxId, id.xPk))),
  };
  return api.register(body) as Promise<{ token: string; mailbox_id: string; region: string }>;
}

export async function login(id: Identity): Promise<string> {
  const ch = await api.challenge();
  const sig = b64e(sign(id.edSk, authMessage(id.mailboxId, ch.nonce)));
  const tok = await api.token({
    mailbox_id: id.mailboxId,
    nonce: ch.nonce,
    signature_b64: sig,
  });
  return tok.token as string;
}

export function makePrekeys(n = 20): { spk: SpkHold; opks: OpkHold[] } {
  const s = freshX();
  const spk = { id: 1, sk: s.sk, pk: s.pk };
  const opks = Array.from({ length: n }, (_, i) => {
    const k = freshX();
    return { id: i + 1, sk: k.sk, pk: k.pk };
  });
  return { spk, opks };
}

export function peerFromCode(code: string): Peer {
  const p = parseBleepCode(code);
  return {
    mailboxId: p.mailboxId,
    edPk: b64e(p.edPk),
    xPk: b64e(p.xPk),
    displayName: p.mailboxId.slice(0, 6),
    verified: false,
    blocked: false,
    accepted: false,
    region: p.region,
  };
}

function getRatchet(snap: VaultSnapshot, peerId: string): RatchetState | null {
  const rec = snap.sessions[peerId];
  if (!rec) return null;
  return loadRatchet(rec.ratchet);
}

function putRatchet(snap: VaultSnapshot, peerId: string, st: RatchetState, extra: Partial<SessionRec> = {}) {
  const prev = snap.sessions[peerId];
  snap.sessions[peerId] = {
    ...prev,
    ...extra,
    peerMailboxId: peerId,
    ratchet: dumpRatchet(st),
    established: extra.established ?? prev?.established ?? true,
  };
}

export async function ensureAliceSession(
  snap: VaultSnapshot,
  token: string,
  peer: Peer,
  spkHold: SpkHold,
): Promise<RatchetState> {
  const existing = getRatchet(snap, peer.mailboxId);
  if (existing && snap.sessions[peer.mailboxId]?.established) return existing;
  const bundle = await api.prekey(peer.mailboxId, token);
  const eph = freshX();
  const bobSpk = b64d(bundle.signed_prekey.pub_b64);
  const bobOpk = bundle.one_time_prekey ? b64d(bundle.one_time_prekey.pub_b64) : null;
  const sk = x3dhAlice(snap.identity.xSk, eph.sk, b64d(peer.xPk), bobSpk, bobOpk);
  const st = initAlice(sk, bobSpk);
  putRatchet(snap, peer.mailboxId, st, {
    established: true,
    pendingEphSk: b64e(eph.sk),
    pendingSpkId: bundle.signed_prekey.key_id,
    pendingOpkId: bundle.one_time_prekey?.key_id,
    theirSpk: bundle.signed_prekey.pub_b64,
  });
  // stash eph pub on the state via a side field on session
  (snap.sessions[peer.mailboxId] as SessionRec & { ephPk?: string }).ephPk = b64e(eph.pk);
  void spkHold;
  return st;
}

export type Payload =
  | { kind: "text"; text: string; id: string; expireSec?: number | null; forwarded?: boolean; viewOnce?: boolean }
  | { kind: "photo" | "voice" | "file"; text: string; id: string; mime: string; name: string; bytesB64: string; expireSec?: number | null; viewOnce?: boolean }
  | {
      kind: "intro";
      text: string;
      displayName: string;
      id: string;
      chamber?: boolean;
      burnAt?: number;
      ephCode?: string;
      groupId?: string;
      groupName?: string;
    }
  | { kind: "react"; id: string; target: string; emoji: string }
  | { kind: "receipt"; id: string; upTo: string }
  | { kind: "typing"; id: string }
  | { kind: "ctrl"; id: string; op: string; data?: unknown }
  | { kind: "call"; id: string; phase: "offer" | "answer" | "ice" | "hangup" | "missed"; sdp?: string; candidate?: string; media?: "audio" | "video" };

export async function sendSealed(
  snap: VaultSnapshot,
  token: string,
  peer: Peer,
  payload: Payload,
  opts: { ttl?: number; spk?: SpkHold; opkMap?: Map<number, OpkHold> } = {},
): Promise<void> {
  const st = await ensureAliceSession(snap, token, peer, opts.spk ?? { id: 1, sk: snap.identity.xSk, pk: snap.identity.xPk });
  const rec = snap.sessions[peer.mailboxId] as SessionRec & { ephPk?: string };
  const { header, ciphertext } = ratchetEncrypt(st, encodePt(payload));
  putRatchet(snap, peer.mailboxId, st, rec);
  const isPk = Boolean(rec.pendingEphSk);
  const inner = signInner(snap.identity, {
    v: 1,
    t: payload.kind === "intro" ? "intro" : isPk ? "pkmsg" : "msg",
    sid: b64e(snap.identity.edPk),
    spk: b64e(snap.identity.xPk),
    mid: snap.identity.mailboxId,
    ek: rec.ephPk,
    used_opk: rec.pendingOpkId,
    used_spk: rec.pendingSpkId,
    h: header,
    ct: b64e(ciphertext),
    ts: Date.now(),
  });
  if (isPk) {
    rec.pendingEphSk = undefined;
    rec.ephPk = undefined;
  }
  const blob = sealOuter(b64d(peer.xPk), encodeInner(inner));
  await api.drop(token, {
    dest_mailbox_id: peer.mailboxId,
    blob_b64: b64e(blob),
    ttl_seconds: opts.ttl ?? 36 * 3600,
  });
}

export type Incoming = {
  payload: Payload;
  fromEd: Uint8Array;
  fromX: Uint8Array;
  innerType: string;
  mailboxId: string;
};

export function openIncoming(
  snap: VaultSnapshot,
  blobB64: string,
  holds: { spk: SpkHold; opks: OpkHold[] },
): Incoming | null {
  let innerBytes: Uint8Array;
  try {
    innerBytes = openOuter(snap.identity.xSk, b64d(blobB64));
  } catch {
    return null;
  }
  const inner = decodeInner(innerBytes);
  if (!verifyInner(inner)) return null;
  const fromEd = b64d(inner.sid);
  const fromX = b64d(inner.spk);
  const peerId = inner.mid || findPeerId(snap, fromEd) || "pending:" + b64e(fromEd).slice(0, 12);

  let st = getRatchet(snap, peerId);
  if (!st) {
    if (inner.t !== "pkmsg" && inner.t !== "intro") return null;
    if (!inner.ek) return null;
    const aliceEph = b64d(inner.ek);
    const opk = inner.used_opk != null ? holds.opks.find((k) => k.id === inner.used_opk) : undefined;
    const sk = x3dhBob(snap.identity.xSk, holds.spk.sk, fromX, aliceEph, opk?.sk ?? null);
    st = initBob(sk, holds.spk.sk, holds.spk.pk);
    snap.sessions[peerId] = {
      peerMailboxId: peerId,
      ratchet: dumpRatchet(st),
      established: true,
    };
  }
  const pt = ratchetDecrypt(st, inner.h, b64d(inner.ct));
  putRatchet(snap, peerId, st);
  const payload = JSON.parse(fromUtf8(pt)) as Payload;
  return { payload, fromEd, fromX, innerType: inner.t, mailboxId: peerId };
}

function findPeerId(snap: VaultSnapshot, edPk: Uint8Array): string | null {
  const want = b64e(edPk);
  for (const p of Object.values(snap.peers)) {
    if (p.edPk === want) return p.mailboxId;
  }
  return null;
}

export function dmThreadId(a: string, b: string): string {
  return ["dm", ...[a, b].sort()].join(":");
}

export function upsertPeer(snap: VaultSnapshot, peer: Peer): void {
  const prev = snap.peers[peer.mailboxId];
  snap.peers[peer.mailboxId] = { ...peer, ...prev, ...peer };
}

export function appendMsg(snap: VaultSnapshot, msg: Message): void {
  const list = snap.messages[msg.threadId] ?? [];
  if (list.some((m) => m.id === msg.id)) return;
  list.push(msg);
  snap.messages[msg.threadId] = list;
  const th = snap.threads[msg.threadId];
  if (th) {
    th.lastAt = msg.at;
    th.lastPreview = msg.deleted ? "Deleted" : msg.kind === "voice" ? "Voice note" : msg.kind === "photo" ? "Photo" : msg.text.slice(0, 80);
    if (!msg.fromMe && !th.muted) th.unread += 1;
  }
}

export function ensureDmThread(snap: VaultSnapshot, peer: Peer): Thread {
  const id = dmThreadId(snap.identity.mailboxId, peer.mailboxId);
  if (!snap.threads[id]) {
    snap.threads[id] = {
      id,
      kind: "dm",
      peerMailboxId: peer.mailboxId,
      title: peer.displayName,
      lastPreview: "",
      lastAt: Date.now(),
      unread: 0,
      pinned: false,
      muted: false,
      archived: false,
      disappearingSec: null,
      receipts: true,
      typing: true,
      draft: "",
    };
  }
  return snap.threads[id];
}

export function guardianForSend(
  snap: VaultSnapshot,
  text: string,
  files: { name: string; mime: string; bytes: Uint8Array }[],
  peer?: Peer,
): Hit[] {
  if (snap.settings.guardianMode === "off") return [];
  return inspectDraft({
    text,
    files,
    peerVerified: peer?.verified ?? false,
    safetyJustChanged: Boolean(peer?.lastSafety && !peer.verified),
  });
}

export function myCode(id: Identity): string {
  return bleepCode(id);
}

export function pairSafety(me: Identity, peer: Peer): string {
  return safetyNumber(me.edPk, b64d(peer.edPk));
}

export { generateIdentity, REGION };

File: android/app/src/main/java/za/bleep/personal/protocol/Session.kt
package za.bleep.personal.protocol

import org.json.JSONObject
import za.bleep.personal.net.Relay

data class Peer(
    val mailboxId: String,
    val edPk: ByteArray,
    val xPk: ByteArray,
    var displayName: String,
    var accepted: Boolean = false,
    var verified: Boolean = false,
    var disappearSec: Int? = null,
    var safetyChanged: Boolean = false,
)

data class ChatMsg(
    val id: String,
    val fromMe: Boolean,
    val text: String,
    val at: Long,
    val peerId: String = "",
    val kind: String = "text",
    val mime: String? = null,
    val bytesB64: String? = null,
    val reacted: String? = null,
    val expiresAt: Long? = null,
)

class Session(private val relay: Relay) {
    lateinit var id: Identity
    lateinit var token: String
    val ready: Boolean get() = this::token.isInitialized

    fun relayBase(url: String) {
        relay.base = url.trim().trimEnd('/')
    }
    lateinit var spk: XPair
    val opks = mutableListOf<Pair<Int, XPair>>()
    val peers = mutableMapOf<String, Peer>()
    val ratchets = mutableMapOf<String, RatchetState>()
    val ephPk = mutableMapOf<String, ByteArray>()
    val pendingPk = mutableSetOf<String>()
    val threads = mutableMapOf<String, MutableList<ChatMsg>>()
    val requests = mutableListOf<ChatMsg>()
    val outbox = mutableListOf<JSONObject>()
    var lastIngest = 0

    fun create(displayName: String) {
        id = Crypto.generateIdentity(displayName)
        spk = Crypto.freshX()
        repeat(8) { i -> opks += (i + 1) to Crypto.freshX() }
        val body = JSONObject()
            .put("mailbox_id", id.mailboxId)
            .put("identity_ed25519_b64", Bytes.b64e(id.edPk))
            .put("identity_x25519_b64", Bytes.b64e(id.xPk))
            .put(
                "signed_prekey",
                JSONObject()
                    .put("key_id", 1)
                    .put("pub_b64", Bytes.b64e(spk.pk))
                    .put("sig_b64", Bytes.b64e(Crypto.sign(id.edSk, Crypto.spkMessage(1, spk.pk)))),
            )
            .put(
                "one_time_prekeys",
                org.json.JSONArray().also { arr ->
                    opks.forEach { (kid, k) ->
                        arr.put(JSONObject().put("key_id", kid).put("pub_b64", Bytes.b64e(k.pk)))
                    }
                },
            )
            .put("registration_sig_b64", Bytes.b64e(Crypto.sign(id.edSk, Crypto.registrationMessage(id.mailboxId, id.xPk))))
        val res = relay.register(body)
        token = res.getString("token")
    }

    fun snapshot(): JSONObject {
        val peersJ = JSONObject()
        peers.forEach { (k, p) ->
            peersJ.put(
                k,
                JSONObject()
                    .put("mailboxId", p.mailboxId)
                    .put("edPk", Bytes.b64e(p.edPk))
                    .put("xPk", Bytes.b64e(p.xPk))
                    .put("displayName", p.displayName)
                    .put("accepted", p.accepted)
                    .put("verified", p.verified)
                    .put("safetyChanged", p.safetyChanged)
                    .put("disappearSec", p.disappearSec ?: JSONObject.NULL),
            )
        }
        val threadsJ = JSONObject()
        threads.forEach { (k, list) ->
            threadsJ.put(k, org.json.JSONArray().also { arr -> list.forEach { arr.put(msgJson(it)) } })
        }
        val ratJ = JSONObject()
        ratchets.forEach { (k, st) -> ratJ.put(k, Ratchet.dump(st)) }
        return JSONObject()
            .put("mailboxId", id.mailboxId)
            .put("displayName", id.displayName)
            .put("edSk", Bytes.b64e(id.edSk))
            .put("edPk", Bytes.b64e(id.edPk))
            .put("xSk", Bytes.b64e(id.xSk))
            .put("xPk", Bytes.b64e(id.xPk))
            .put("spkSk", Bytes.b64e(spk.sk))
            .put("spkPk", Bytes.b64e(spk.pk))
            .put("peers", peersJ)
            .put("threads", threadsJ)
            .put("ratchets", ratJ)
            .put("requests", org.json.JSONArray().also { arr -> requests.forEach { arr.put(msgJson(it)) } })
            .put("outbox", org.json.JSONArray().also { arr -> outbox.forEach { arr.put(it) } })
    }

    fun restore(snap: JSONObject) {
        id = Identity(
            snap.getString("mailboxId"),
            Bytes.b64d(snap.getString("edSk")),
            Bytes.b64d(snap.getString("edPk")),
            Bytes.b64d(snap.getString("xSk")),
            Bytes.b64d(snap.getString("xPk")),
            snap.getString("displayName"),
        )
        spk = XPair(Bytes.b64d(snap.getString("spkSk")), Bytes.b64d(snap.getString("spkPk")))
        peers.clear(); threads.clear(); ratchets.clear(); requests.clear()
        snap.optJSONObject("peers")?.keys()?.forEach { k ->
            val o = snap.getJSONObject("peers").getJSONObject(k)
            peers[k] = Peer(
                o.getString("mailboxId"),
                Bytes.b64d(o.getString("edPk")),
                Bytes.b64d(o.getString("xPk")),
                o.getString("displayName"),
                o.optBoolean("accepted"),
                o.optBoolean("verified"),
                if (o.isNull("disappearSec")) null else o.optInt("disappearSec"),
                o.optBoolean("safetyChanged"),
            )
        }
        snap.optJSONObject("threads")?.keys()?.forEach { k ->
            val arr = snap.getJSONObject("threads").getJSONArray(k)
            threads[k] = MutableList(arr.length()) { i -> msgFrom(arr.getJSONObject(i)) }
        }
        snap.optJSONObject("ratchets")?.keys()?.forEach { k ->
            ratchets[k] = Ratchet.load(snap.getJSONObject("ratchets").getJSONObject(k))
        }
        snap.optJSONArray("requests")?.let { arr ->
            for (i in 0 until arr.length()) requests += msgFrom(arr.getJSONObject(i))
        }
        snap.optJSONArray("outbox")?.let { arr ->
            outbox.clear()
            for (i in 0 until arr.length()) outbox += arr.getJSONObject(i)
        }
        val ch = relay.challenge()
        val sig = Crypto.sign(id.edSk, Crypto.authMessage(id.mailboxId, ch.getString("nonce")))
        val tok = relay.token(
            JSONObject()
                .put("mailbox_id", id.mailboxId)
                .put("nonce", ch.getString("nonce"))
                .put("signature_b64", Bytes.b64e(sig)),
        )
        token = tok.getString("token")
    }

    private fun msgJson(m: ChatMsg) = JSONObject()
        .put("id", m.id)
        .put("fromMe", m.fromMe)
        .put("text", m.text)
        .put("at", m.at)
        .put("peerId", m.peerId)
        .put("kind", m.kind)
        .put("mime", m.mime ?: JSONObject.NULL)
        .put("bytesB64", m.bytesB64 ?: JSONObject.NULL)
        .put("reacted", m.reacted ?: JSONObject.NULL)
        .put("expiresAt", m.expiresAt ?: JSONObject.NULL)

    private fun msgFrom(o: JSONObject) = ChatMsg(
        o.getString("id"),
        o.getBoolean("fromMe"),
        o.optString("text"),
        o.getLong("at"),
        o.optString("peerId"),
        o.optString("kind", "text"),
        if (o.isNull("mime")) null else o.optString("mime"),
        if (o.isNull("bytesB64")) null else o.optString("bytesB64"),
        if (o.isNull("reacted")) null else o.optString("reacted"),
        if (o.isNull("expiresAt")) null else o.optLong("expiresAt"),
    )

    fun addByCode(code: String): Peer {
        val (mid, ed, x) = Crypto.parseCode(code)
        val p = Peer(mid, ed, x, mid.take(6))
        peers[mid] = p
        return p
    }

    fun sendText(peer: Peer, text: String) {
        if (peer.safetyChanged) error("Safety number changed. This is a new key. Verify it before you send.")
        ensureAlice(peer)
        val st = ratchets[peer.mailboxId]!!
        val exp = peer.disappearSec?.let { System.currentTimeMillis() + it * 1000L }
        val payload = JSONObject().put("kind", "text").put("id", java.util.UUID.randomUUID().toString()).put("text", text)
        if (peer.disappearSec != null) payload.put("expireSec", peer.disappearSec)
        val (header, ct) = Ratchet.encrypt(st, Bytes.utf8(payload.toString()))
        val isPk = pendingPk.remove(peer.mailboxId)
        val unsigned = JSONObject()
            .put("v", 1)
            .put("t", if (isPk) "pkmsg" else "msg")
            .put("sid", Bytes.b64e(id.edPk))
            .put("spk", Bytes.b64e(id.xPk))
            .put("mid", id.mailboxId)
            .put("h", header.json())
            .put("ct", Bytes.b64e(ct))
            .put("ts", System.currentTimeMillis())
        ephPk.remove(peer.mailboxId)?.let { unsigned.put("ek", Bytes.b64e(it)) }
        val sig = Crypto.sign(id.edSk, Bytes.utf8(Canonical.of(unsigned)))
        unsigned.put("sig", Bytes.b64e(sig))
        val blob = Seal.outer(peer.xPk, Bytes.utf8(unsigned.toString()))
        enqueueDrop(peer.mailboxId, Bytes.b64e(blob))
        val list = threads.getOrPut(peer.mailboxId) { mutableListOf() }
        list += ChatMsg(payload.getString("id"), true, text, System.currentTimeMillis(), peer.mailboxId, expiresAt = exp)
    }

    fun sendReact(peer: Peer, targetId: String, emoji: String = "👍") {
        sendPayload(peer, JSONObject().put("kind", "react").put("id", java.util.UUID.randomUUID().toString()).put("target", targetId).put("emoji", emoji), intro = false)
        val list = threads[peer.mailboxId] ?: return
        val i = list.indexOfFirst { it.id == targetId }
        if (i >= 0) list[i] = list[i].copy(reacted = emoji)
    }

    fun sweep() {
        val now = System.currentTimeMillis()
        threads.values.forEach { list ->
            list.replaceAll { m ->
                if (m.expiresAt != null && m.expiresAt < now) m.copy(text = "", kind = "system") else m
            }
        }
    }

    fun sendMedia(peer: Peer, kind: String, bytes: ByteArray, mime: String, name: String) {
        if (peer.safetyChanged) error("Safety number changed. This is a new key. Verify it before you send.")
        val id = java.util.UUID.randomUUID().toString()
        val payload = JSONObject()
            .put("kind", kind)
            .put("id", id)
            .put("text", "")
            .put("mime", mime)
            .put("name", name)
            .put("bytesB64", Bytes.b64e(bytes))
        sendPayload(peer, payload, intro = false)
        threads.getOrPut(peer.mailboxId) { mutableListOf() } +=
            ChatMsg(id, true, if (kind == "voice") "Voice note" else "Photo", System.currentTimeMillis(), peer.mailboxId, kind, mime, Bytes.b64e(bytes))
    }

    fun sendIntro(peer: Peer, hello: String) {
        sendPayload(peer, JSONObject().put("kind", "intro").put("id", java.util.UUID.randomUUID().toString()).put("text", hello).put("displayName", id.displayName), intro = true)
    }

    private fun sendPayload(peer: Peer, payload: JSONObject, intro: Boolean) {
        ensureAlice(peer)
        val st = ratchets[peer.mailboxId]!!
        val (header, ct) = Ratchet.encrypt(st, Bytes.utf8(payload.toString()))
        val isPk = pendingPk.remove(peer.mailboxId)
        val unsigned = JSONObject()
            .put("v", 1)
            .put("t", if (intro) "intro" else if (isPk) "pkmsg" else "msg")
            .put("sid", Bytes.b64e(id.edPk))
            .put("spk", Bytes.b64e(id.xPk))
            .put("mid", id.mailboxId)
            .put("h", header.json())
            .put("ct", Bytes.b64e(ct))
            .put("ts", System.currentTimeMillis())
        ephPk.remove(peer.mailboxId)?.let { unsigned.put("ek", Bytes.b64e(it)) }
        unsigned.put("sig", Bytes.b64e(Crypto.sign(id.edSk, Bytes.utf8(Canonical.of(unsigned)))))
        enqueueDrop(peer.mailboxId, Bytes.b64e(Seal.outer(peer.xPk, Bytes.utf8(unsigned.toString()))))
    }

    private fun enqueueDrop(dest: String, blobB64: String) {
        try {
            relay.drop(token, dest, blobB64)
        } catch (_: Exception) {
            outbox += JSONObject().put("dest", dest).put("blob", blobB64)
        }
    }

    fun flushOutbox() {
        val left = mutableListOf<JSONObject>()
        for (q in outbox) {
            try {
                relay.drop(token, q.getString("dest"), q.getString("blob"))
            } catch (_: Exception) {
                left += q
            }
        }
        outbox.clear()
        outbox.addAll(left)
    }

    fun poll() {
        lastIngest = 0
        flushOutbox()
        val items = relay.fetch(token)
        for (i in 0 until items.length()) {
            val blob = items.getJSONObject(i).getString("blob_b64")
            ingest(blob)
        }
        sweep()
    }

    private fun ingest(blobB64: String) {
        val innerBytes = try {
            Seal.openOuter(id.xSk, Bytes.b64d(blobB64))
        } catch (_: Exception) {
            return
        }
        val inner = JSONObject(Bytes.fromUtf8(innerBytes))
        val rest = JSONObject(inner.toString()).apply { remove("sig") }
        if (!Crypto.verify(Bytes.b64d(inner.getString("sid")), Bytes.utf8(Canonical.of(rest)), Bytes.b64d(inner.getString("sig")))) return
        val mid = inner.getString("mid")
        val fromEd = Bytes.b64d(inner.getString("sid"))
        val fromX = Bytes.b64d(inner.getString("spk"))
        peers[mid]?.let { existing ->
            if (!existing.edPk.contentEquals(fromEd)) {
                existing.safetyChanged = true
                existing.verified = false
            }
        }
        var st = ratchets[mid]
        if (st == null) {
            if (!inner.has("ek")) return
            val sk = Ratchet.x3dhBob(id.xSk, spk.sk, fromX, Bytes.b64d(inner.getString("ek")), null)
            st = Ratchet.initBob(sk, spk.sk, spk.pk)
            ratchets[mid] = st
        }
        val h = inner.getJSONObject("h")
        val header = Header(h.getString("dh"), h.getInt("n"), h.getInt("pn"))
        val pt = JSONObject(Bytes.fromUtf8(Ratchet.decrypt(st, header, Bytes.b64d(inner.getString("ct")))))
        val kind = pt.optString("kind")
        if (kind == "react") {
            val target = pt.optString("target")
            val emoji = pt.optString("emoji")
            threads[mid]?.let { list ->
                val i = list.indexOfFirst { it.id == target }
                if (i >= 0) list[i] = list[i].copy(reacted = emoji)
            }
            return
        }
        val text = pt.optString("text")
        val peer = peers.getOrPut(mid) {
            Peer(mid, Bytes.b64d(inner.getString("sid")), fromX, mid.take(6), accepted = false)
        }
        val bytesB64 = pt.optString("bytesB64").ifBlank { null }
        val msgKind = if (kind in listOf("photo", "voice", "file")) kind else "text"
        val label = when (msgKind) {
            "voice" -> "Voice note"
            "photo" -> "Photo"
            else -> text.ifBlank { "Request to message" }
        }
        val msg = ChatMsg(pt.optString("id"), false, label, System.currentTimeMillis(), mid, msgKind, pt.optString("mime").ifBlank { null }, bytesB64)
        if (!peer.accepted || kind == "intro") {
            requests += msg
            lastIngest++
            if (kind == "intro" && pt.has("displayName")) peer.displayName = pt.getString("displayName")
            return
        }
        threads.getOrPut(mid) { mutableListOf() } += msg
        lastIngest++
    }

    fun verifyPeer(peer: Peer) {
        peer.verified = true
        peer.safetyChanged = false
    }

    fun unlink() {
        if (ready) relay.unlink(id.mailboxId, token)
    }

    fun accept(mailboxId: String) {
        peers[mailboxId]?.accepted = true
        val moved = requests.filter { it.peerId == mailboxId || it.peerId.isEmpty() }
        requests.removeAll(moved.toSet())
        threads.getOrPut(mailboxId) { mutableListOf() }.addAll(moved.map { it.copy(peerId = mailboxId) })
    }

    private fun ensureAlice(peer: Peer) {
        if (ratchets[peer.mailboxId] != null) return
        val bundle = relay.prekey(peer.mailboxId, token)
        val eph = Crypto.freshX()
        val bobSpk = Bytes.b64d(bundle.getJSONObject("signed_prekey").getString("pub_b64"))
        val opk = bundle.optJSONObject("one_time_prekey")?.let { Bytes.b64d(it.getString("pub_b64")) }
        val sk = Ratchet.x3dhAlice(id.xSk, eph.sk, peer.xPk, bobSpk, opk)
        ratchets[peer.mailboxId] = Ratchet.initAlice(sk, bobSpk)
        ephPk[peer.mailboxId] = eph.pk
        pendingPk += peer.mailboxId
    }
}
